### DISCUSSION TITLE
issue: shouldUseNativeValidation and novalidate not working

### DISCUSSION THREAD

**Discussion by jrafaaael** [react-hook-form/react-hook-form#7463] [Q&A]
### Version Number

7.22.5

### Codesandbox/Expo snack

https://codesandbox.io/s/red-breeze-lqnt3

### Steps to reproduce

1. Provide an invalid value in any input field
2. Submit the input
3. Edit the input with a invalid value, again
4. See the bug
![Screenshot at 2022-01-06 18-30-03](https://user-images.githubusercontent.com/70046023/148461398-83c05997-e624-467d-aa09-9c25ee06e0c0.png)

![Uploading Screenshot at 2022-01-06 18-30-03.png…]()

### Expected behaviour

The native validation should not appear.
Also, in the initial state, the input takes it as invalid (which is correct because it is required and at that moment it does not have any character) but the validation should occur after firing the onSubmit event and not before

### What browsers are you seeing the problem on?

Firefox, Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
If you set `shouldUseNativeValidation` to true then you are choose to use native validation, or just leave that config out which is the default.

**Reply by jrafaaael** [reply]
From documentacion:

> This config will enable browser native validation. It will also enable CSS selectors :valid and:invalid making style inputs easier. In fact, you can still use those selectors even the client validation is disabled.
> -  Note: You can turn on this config and set `novalidate` at your form and still use those CSS selectors.
```JS 
  // you can still disabled the native validation, CSS selectors such as
  // invalid and valid still going to work 
  // <form onSubmit={handleSubmit(onSubmit)} novalidate>
```

I'm trying this but not working (see screenshots and codesandbox)

**Reply by jrafaaael** [reply]
Sorry, the ss not loaded in the issue:

> The validation occurs before onSubmit event (see the red border of the input indicating a invalid entry. This style only is applied when `.input:invalid` occurs)

![Screenshot at 2022-01-06 18-41-36](https://user-images.githubusercontent.com/70046023/148463666-52deec98-3f4a-404f-8082-e276c1dfd48e.png)


> The validation still occurs before onSubmit event (see the blue border of the input indicating a valid entry. This style is applied by default)

![Screenshot at 2022-01-06 18-42-02](https://user-images.githubusercontent.com/70046023/148463700-148016b9-1836-4f09-b435-456bc677f2d2.png)


> Native validation occurs after onSubmit fires and the input change to a invalid value

![Screenshot at 2022-01-06 18-43-54](https://user-images.githubusercontent.com/70046023/148463709-d5c0833d-10a7-4eaf-ad80-52cf31f3bd97.png)

**Reply by bluebill1049** [reply]
going to update the doc as it's misleading.

**Comment by bluebill1049** [comment]
> Also, in the initial state, the input takes it as invalid (which is correct because it is required and at that moment it does not have any character) but the validation should occur after firing the onSubmit event and not before

provide a isolated example with steps to reproduce.

**Comment by psirenny** [comment]
The issue is that inputs are marked as :valid or :invalid before the form is submitted. Therefore, new empty forms show each input as invalid. The `novalidate` attribute is a red herring.

**Comment by mmalomo** [comment]
Facing the same problem. 
`shouldUseNativeValidation` and `novalidate` are not working as it is said in the documentation.

I think the problem is [here](https://github.com/react-hook-form/react-hook-form/blob/dd547079dc12b55fa44bd0135a94a8f5d595dfcc/src/logic/validateField.ts#L52). If `shouldUseNativeValidation` is true, `setCustomValidity` is called and the browser ignores the `novalidate` attribute in the form.

Maybe the `setCustomValidity` function must check if the form has the novalidate attribute? @bluebill1049 @jrafaaael

**Reply by bluebill1049** [reply]
Curious the reason to use that prop if u want to disable build in validator?

**Reply by thiago-silva-ac** [reply]
Same problem here, my case is that's I just want the easy css selectors for :valid and :invalid fields.

**Reply by eduards** [reply]
I'm facing the same problem, just for another reason: 

I set `shouldUseNativeValidation: true` so that react-hook-form sets the `required` attribute on the input HTML to make sure that screen readers announce this to the user. Without setting `shouldUseNativeValidation` to true, neither the `required` nor the `aria-required`/`aria-invalid`/`aria-errormessage` attributes are set on the input element. In addition to that, I set the form's `novalidate` attribute to true as I don't want the native validation behaviour. 

I did expect that the behaviour matches the behaviour of a native form element whose `novalidate` attribute is set to true. However, this isn't the case. The native validation error messages are still reported to the user as already described above.
